### DISCUSSION TITLE
`State` struct

### DISCUSSION THREAD

**Discussion by longemen3000** [ClapeyronThermo/Clapeyron.jl#269] [General]
with #268 , the need for  `State` struct seems to arise. this discussion is to discuss the details of a `State` struct for Clapeyron.
- What is a `State` struct? : it is a struct that contains a set of thermodynamic properties that constitute a complete state. For example: T = 300K, p = 1 atm, z = [0.5,0.5] is a complete state for a 2-component model. there are also states that depend on the phase (liquid, gas, solid) or equilibria conditions (at saturation, critical point)
- what interface do we want for the `State` constructor?
- how do we calculate properties with the `State` constructor?
- compatibility

**Comment by longemen3000** [comment]
Coolprop supports the following:
Input Pair          |Description
:-------------------|:------------------------------------------------
QT_INPUTS           |Molar quality, Temperature in K
QSmolar_INPUTS      |Molar quality, Entropy in J/mol/K
QSmass_INPUTS       |Molar quality, Entropy in J/kg/K
HmolarQ_INPUTS      |Enthalpy in J/mol, Molar quality
HmassQ_INPUTS       |Enthalpy in J/kg, Molar quality
DmassQ_INPUTS       |Molar density kg/m^3, Molar quality
DmolarQ_INPUTS      |Molar density in mol/m^3, Molar quality
PQ_INPUTS           |Pressure in Pa, Molar quality
PT_INPUTS           |Pressure in Pa, Temperature in K
DmassT_INPUTS       |Mass density in kg/m^3, Temperature in K
DmolarT_INPUTS      |Molar density in mol/m^3, Temperature in K
HmassT_INPUTS       |Enthalpy in J/kg, Temperature in K
HmolarT_INPUTS      |Enthalpy in J/mol, Temperature in K
SmassT_INPUTS       |Entropy in J/kg/K, Temperature in K
SmolarT_INPUTS      |Entropy in J/mol/K, Temperature in K
TUmass_INPUTS       |Temperature in K, Internal energy in J/kg
TUmolar_INPUTS      |Temperature in K, Internal energy in J/mol
DmassP_INPUTS       |Mass density in kg/m^3, Pressure in Pa
DmolarP_INPUTS      |Molar density in mol/m^3, Pressure in Pa
HmassP_INPUTS       |Enthalpy in J/kg, Pressure in Pa
HmolarP_INPUTS      |Enthalpy in J/mol, Pressure in Pa
PSmass_INPUTS       |Pressure in Pa, Entropy in J/kg/K
PSmolar_INPUTS      |Pressure in Pa, Entropy in J/mol/K
PUmass_INPUTS       |Pressure in Pa, Internal energy in J/kg
PUmolar_INPUTS      |Pressure in Pa, Internal energy in J/mol
DmassHmass_INPUTS   |Mass density in kg/m^3, Enthalpy in J/kg
DmolarHmolar_INPUTS |Molar density in mol/m^3, Enthalpy in J/mol
DmassSmass_INPUTS   |Mass density in kg/m^3, Entropy in J/kg/K
DmolarSmolar_INPUTS |Molar density in mol/m^3, Entropy in J/mol/K
DmassUmass_INPUTS   |Mass density in kg/m^3, Internal energy in J/kg
DmolarUmolar_INPUTS |Molar density in mol/m^3, Internal energy in J/mol
HmassSmass_INPUTS   |Enthalpy in J/kg, Entropy in J/kg/K
HmolarSmolar_INPUTS |Enthalpy in J/mol, Entropy in J/mol/K
SmassUmass_INPUTS   |Entropy in J/kg/K, Internal energy in J/kg
SmolarUmolar_INPUTS |Entropy in J/mol/K, Internal energy in J/mol

**Reply by pw0908** [reply]
This might be a bit excessive, but we could create an extension for CoolProp.jl? We thought (long ago) about the `State` approach, but decided against it as, when generating phase diagrams, it's far more convenient to have these query functions. However, to be compatible with process simulation packages, like Ai4EComponentsLib.jl, being compatible with CoolProp.jl will go a long way.

**Reply by longemen3000** [reply]
Yeah, we can do both, but we need the state dispatcher to properties, with a state struct, we can build a Coolprop.jl extension with ease

**Reply by pw0908** [reply]
I'm a bit confused at to how these work. Do they assume you've also specified molar compositions? Cause otherwise the state is under specified.

**Reply by longemen3000** [reply]
Yes, a Coolprop model is actually a model + molar compositions. The low level interface of Coolprop is more akin to how Clapeyron works. In the low level api, first you create a model, then set the compositions, resolve the state, and finally calculate the desired property. We want to perform the state resolver ourselves.

**Reply by pw0908** [reply]
So our implementation would be composition + any of those two variables?

**Reply by pw0908** [reply]
In that case, then we can define a state as:

```julia
struct State
   temperature::Float64
   pressure::Float64
   volume::Float64
   composition::Vector{Float64}
   internal_energy::Float64
   etc...
end

function State(z::Vector{Float64}; temperature::Float64 = nothing, pressure::Float64 = nothing, etc...)

   if check specified variables if supported

   else 
      error
   end

   return state
end
```

We can force composition to be specified.

**Reply by longemen3000** [reply]
If you add the model in that struct and make it mutable, you basically have the same structure that CoolProp uses internally

**Reply by pw0908** [reply]
I see. That should work then?

Apart from that, would we now need a VT-solver?

**Reply by pw0908** [reply]
Oh! We could also include caches at this point!

**Reply by longemen3000** [reply]
Yeah, that is an option too

**Reply by pw0908** [reply]
So do we want to figure out this `State` framework before the next release?

**Reply by longemen3000** [reply]
I'm interested. At the moment I'm finishing the COSMO updates, after that I can work up on that

**Comment by Sush1090** [comment]
Hello,
I came across this discussion recently. 
There is something similar I do locally. It is same as the `State` described here. 

It has the following vars
```julia
mutable struct State
    model ::EoSModel
    z:: AbstractVector
    p::Any 
    T::Any 
    h::Any 
    s::Any
    v::Any 
    q::Any 
    method::Symbol
end
```

I store the `method` as a symbol to choose which flash computation it should follow

For example I can do the following:

```julia
model = cPR(["pentane","isobutane"],idealmodel = ReidIdeal)
z = [1.0,1.0]; p = 101325*3; T = 350; h = Clapeyron.enthalpy(model,p,T,z)

S = State(model=model,z= z,p = p,h = h)
```
Here I define the `State` to be a `ph` flash based solution. Then I use that to find the property I need for example :
```julia
julia> entropy(S)
-49.84418964661152
```
I would be glad to contribute to this if it is still of interest. 

Thank you,
Sushrut

**Comment by longemen3000** [comment]
Hello,

This was originally a discussion on how to support multiple inputs, apart from PT, when i wrote this, i just didn't have any idea on how to implement those.

Now, we have a sort of structure on which we can work with:
- all (supported) solvers have a `XY_property` function
- all flashes return a `FlashResult.`
- all (supported) properties allow calling `f(model, state)`

`FlashResult` assumed the role i wanted for `State` to be, at the moment.

With that said, what do you have in mind? a way to call a flash with the keyword arguments defining the type of flash?

**Comment by Sush1090** [comment]
I had made this as a user friendly (just for myself haha) way to go through all the various Flash computations (`pt`,`ph` etc ...) 

I was looking for something like `PropsSI("output","var1"val1,"var2"val2,fluid)` where I do not specify explcilty on which flash needs to be used. It will be automatic. As of now we still need to import the relevant `xy_flash` to do our computation. 

> With that said, what do you have in mind? a way to call a flash with the keyword arguments defining the type of flash?

Yes exactly as we define the `State` only once.

**Reply by Sush1090** [reply]
In other words its just a wrapper over all the flash computations for ease of usage

**Reply by longemen3000** [reply]
Yeah, that would be great to have, something like `Clapeyron.flash(model;kwargs...)`?

**Reply by Sush1090** [reply]
yes indeed. As of now my code is very "dirty". I can clean it up and make a pull request

**Reply by longemen3000** [reply]
Seems good to have, feel free to open up a PR
